### PR TITLE
New signing interface

### PR DESCRIPTION
**Pull request by ltseeley**

### CODE REVIEW

File: libcylinder/src/key.rs [hunk 1]
```diff
@@ -0,0 +1,111 @@
+/*
```
Comment (agunde406) [inline libcylinder/src/key.rs:1]: I think you are missing a word in your commit message. Should be "These structs will be used throughout the library."

File: libcylinder/src/lib.rs [hunk 2]
```diff
@@ -15,5 +16,30 @@
...
+pub use error::{SignatureVerificationError, SigningError};
+pub use key::{PrivateKey, PublicKey};
+
+/// A signer for arbitrary messages
+pub trait Signer: Send {
+    /// Signs the given message
+    fn sign(&self, message: &[u8]) -> Result<Vec<u8>, SigningError>;
+
+    /// Returns the signer's public key
+    fn public_key(&self) -> Result<PublicKey, SigningError>;
```
Comment (vaporos) [inline libcylinder/src/lib.rs:33]: This should return Result<&PublicKey, SIgningError> so we aren't forcing a clone. The first time we generate the key, we should cache it within the Signer and return the same thing if multiple public_key() calls are made.

File: libcylinder/src/key.rs (hunk 1 above)
Comment (ltseeley) [reply libcylinder/src/key.rs:1]: Fixed

File: libcylinder/src/lib.rs (hunk 2 above)
Comment (ltseeley) [reply libcylinder/src/lib.rs]: Updated

Comment (peterschwarz) [reply libcylinder/src/lib.rs:33]: This is actually quite hard to do in practice, given how safe interior mutability works.  It is probably doable, but we'd be forcing all implementations to use unsafe interior mutability.

**Pull request merged**